### PR TITLE
lint: lint snap files inside an instance

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [X] Have you successfully run `make lint`?
- [X] Have you successfully run `pytest tests/unit`?

-----

So far, `snapcraft lint` parses the command line, prepares an instance, and re-executes snapcraft inside the instance.

This PR adds the code that is executed inside the instance:
1. Unsquash the snap
2. Collect the metadata from snap.yaml and lint filters from snapcraft.yaml
3. Install the snap
4. Run the linter on `/snap/<snap-name>/current`
5. Report the linter warnings

(CRAFT-1689)

### CODE REVIEW

File: tests/unit/commands/test_lint.py [hunk 1]
```diff
@@ -138,114 +209,719 @@ def test_lint_http_https_proxy(
...
+    emitter.assert_debug(f"Could not find {snapcraft_yaml_file.name!r}.")
+
+
+@pytest.mark.parametrize("base", ["core", "core18", "core20"])
+def test_load_project_unsupported_core_error(base, tmp_path):
+    """Raise an error if for snaps with core, core18, and core20 bases."""
+    # create a simple snapcraft.yaml
+    (tmp_path / "snap").mkdir()
+    snap_file = tmp_path / "snap/snapcraft.yaml"
+    with snap_file.open("w") as yaml_file:
```
Comment (tigarmo) [inline tests/unit/commands/test_lint.py:904]: it's a (very) minor suggestion but `yaml_file.write_text(dedent(...))` might look cleaner

File: snapcraft/commands/lint.py [hunk 2]
```diff
@@ -177,18 +194,189 @@ def _prepare_instance(
...
+            ) from error
+
+        return Path("/snap") / snap_metadata.name / "current"
+
+    def _load_lint_filters(self, project: Optional[projects.Project]) -> projects.Lint:
+        """Load lint filters from a Project and disable the classic linter.
+
+        :param project: Project from the snap file, if present.
+
+        :returns: Lint config with classic linter disabled.
```
Comment (tigarmo) [inline snapcraft/commands/lint.py:352]: out of curiosity, can you tell me why the classic linter is disabled?

**Review by tigarmo — APPROVED** [review]
Callahan I'm afraid I tried my best to find issues with your code but I failed :/
this is very clear and well-organized, very nice! The unit tests are verbose with the mocking but it's consistent with the other tests, so no issues there.
thanks!

File: tests/unit/commands/test_lint.py (hunk 1 above)
Comment (mr-cal) [reply tests/unit/commands/test_lint.py]: Good suggestion!

**Pull request merged**